### PR TITLE
Initial version of the libcephfs proxy

### PR DESCRIPTION
**Pull request by xhernandez**
The libcephfs proxy is composed by a new daemon process (libcephfsd) and a new library (libcephfs_proxy.so) that allow client applications to work without directly connecting to the libcephfs.so library.

The libcephfs_proxy.so library exports binary compatible versions of the libcephfs.so functions. These functions forward the application requests to the libcephfsd daemon running on the same machine through a UNIX socket. The daemon, which is linked to the real libcephfs.so library, executes the incoming requests and sends back the answers.

The client application doesn't need to be modified to use the proxy. It just needs to link against libcephfs_proxy.so instead of libcephfs.so, or even use LD_PRELOAD when executing it.

The main advantage of using the proxy is that all ceph mounts are managed in a single place (the libcephfsd daemon), which will help reduce memory usage when many processes are connecting to the same Ceph volumes using the same configuration.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by xhernandez** [comment]
This PR doesn't implement the full set of functions from libcephfs, but provides enough functionality to support Samba's VFS Ceph module.

For now it uses a thread per connection. I'm working on an epoll-based approach to handle all connections from a smaller number of threads.

**Comment by xhernandez** [comment]
> Just a heads up - I noticed that there's changes to cmake config but not to the packages (rpm & deb). I'm not sure about deb packaging but I think there might be a build failure if the rpm detects unpackaged files in the build. If you can maybe try a test build of the packages locally and see if it might disrupt the build jobs if merged as is - and if so, let include packaging update patches in the PR too.

I've verified that it fails building RPM packages, so I've added build instructions for rpm and deb, but I don't have experience building packages, so I hope it's fine. Let me know if you find any issues.

**Comment by phlogistonjohn** [comment]
> I've verified that it fails building RPM packages, so I've added build instructions for rpm and deb, but I don't have experience building packages, so I hope it's fine. Let me know if you find any issues

The issue is that the ceph integration test process involves building packages. So let's say that we were to merge the PR with no changes to the spec file. Suddenly, main and any branch inheriting from main will fail builds and nobody can do teuthology testing until it's fixed. Therefore, I think having patches in the PR to update the packging files, in addition to the cmake files, is a blocker.

That said I'm willing to help on this front, at least with regards to the spec file. Maybe @anoopcs9 would be able to help here as well. We may need to ask around for assistance with regards to the deb packages.

One other option would be to make this component conditional and default to off. That way you have to explicitly ask cmake to build the proxy. In that case we could merge code without it being visible to the packaging intially. Then in a follow up pr we can set the build options in the spec and update the spec for the new files generated. 

I have no strong opinions either way - I just want to avoid one of those "teuthology is totally broken for everybody for a day or two" type situations.

**Comment by phlogistonjohn** [comment]
Er, ignore everything I just wrote. When you said "instructions" in your other comment I thought you mean actual "text for a human to follow". I'm sorry to say I didn't look at the patches first. It looks like a good start and when we're ready we can push to the ci repo and test the builds with these patches. Sorry for the noise!

File: debian/control
Comment (anthonyeleven) [inline debian/control:934]: CephFS mounts

File: debian/control
Comment (anthonyeleven) [inline debian/control:946]: Ditto

File: doc/dev/libcephfs_proxy.rst
Comment (anthonyeleven) [inline doc/dev/libcephfs_proxy.rst:10]: S/eat/consume/

File: debian/control
Comment (xhernandez) [reply debian/control]: Done

File: debian/control
Comment (xhernandez) [reply debian/control]: DOne

File: doc/dev/libcephfs_proxy.rst
Comment (xhernandez) [reply doc/dev/libcephfs_proxy.rst]: Done

**Comment by mchangir** [comment]
still going through the PR ...
but, I'd rather not have an identifier named `main` other than the default process entry point

File: src/proxy/proxy_manager.c
Comment (mchangir) [inline src/proxy/proxy_manager.c:230]: if possible, please rename identifier `main` to something more specific

File: src/proxy/proxy_manager.c
Comment (xhernandez) [reply src/proxy/proxy_manager.c]: Done

File: src/proxy/proxy_manager.c
Comment (xhernandez) [reply src/proxy/proxy_manager.c]: Done

**Comment by xhernandez** [comment]
I've addressed the review comments and fixed some issues we detected during testing.

File: src/proxy/CMakeLists.txt
Comment (vshankar) [inline src/proxy/CMakeLists.txt:1]: nit: lets keep the path as `src/libcephfs_proxy`?

**Review by vshankar — COMMENTED** [review]
I haven't done an in-depth review of the change @xhernandez, but some tests would really help, especially since the proxy is pretty much pluggable and required no changes to applications currently using libcephfs.

File: doc/dev/libcephfs_proxy.rst
Comment (xhernandez) [reply doc/dev/libcephfs_proxy.rst:14]: The reconfigurability of libcephfs.so is already stated in the previous sentence. Here it talks about using multiple independent instances of libcephfs, and in this configuration it's pretty easy to trigger the OOM killer. Even configuring the memory limits of each instance it's not possible to avoid consuming all the memory since the number of instances is unbounded.

Anyway, I'll rewrite it to try to explain it in more detail.

**Comment by xhernandez** [comment]
> I haven't done an in-depth review of the change @xhernandez, but some tests would really help, especially since the proxy is pretty much pluggable and required no changes to applications currently using libcephfs.

Right now, the proxy only implements a subset of the functions from libcephfs. This is because it has been initially implemented to try to reduce the resource consumption when used from Samba. For this reason, this initial implementation only provides the subset of functions required by the Samba's Ceph low-level VFS module.

This means that switching existing tests from using _libcephfs.so_ to _libcephfs_proxy.so_ won't work in some (many ?) cases for now. However, Samba already has an extensive test suite, and the proxy is already being (indirectly) tested when all those tests go through the Ceph VFS module. I think this should be a good enough for now.

Eventually all libcephfs functions should be implemented. When that happens we could execute the same set of tests currently run against _libcephfs.so_ on the new proxy, and have the same coverage.

File: doc/dev/libcephfs_proxy.rst
Comment (xhernandez) [reply doc/dev/libcephfs_proxy.rst:84]: Done

File: ceph.spec.in
Comment (spuiuk) [inline ceph.spec.in:2540]: The binary is found in the sbin directory.  This line should be
%{_sbindir}/libcephfsd

instead

**Review by spuiuk — CHANGES_REQUESTED** [review]
The following change is required for a successful rpm build.

File: ceph.spec.in
Comment (xhernandez) [reply ceph.spec.in]: Done

**Comment by anthonyeleven** [comment]
I thought CephFS mounts are done by the kernel and thus access is coalesced that way.  Am I abjectly mistaken?

**Comment by xhernandez** [comment]
> I thought CephFS mounts are done by the kernel and thus access is coalesced that way. Am I abjectly mistaken?

A kernel mount is one way to access a CephFS filesystem. Another way is by using the _libcephfs.so_ library, which provides functions to access the same CephFS filesystem but without mounting it through the kernel. It has some advantages in some cases, like not requiring root permissions to access the filesystem (only root can do a kernel mount), or be able to do some operations directly to the inode.

**Comment by anthonyeleven** [comment]
Thanks! Now I know.

**Comment by vshankar** [comment]
jenkins test api

**Comment by slowfranklin** [comment]
Thinking about it, is there a way to do the metadata operations (including possibly open()) via the proxy, to benefit from (presumably) the shared cache and reduced memory footprint, but do IO (read/write) directly, bypassing the proxy? I imagine that might give you the best of both worlds, sort of...

**Comment by xhernandez** [comment]
> Thinking about it, is there a way to do the metadata operations (including possibly open()) via the proxy, to benefit from (presumably) the shared cache and reduced memory footprint, but do IO (read/write) directly, bypassing the proxy? I imagine that might give you the best of both worlds, sort of...

If we want to bypass the proxy for some requests, read() and write() are not the best candidates. In some tests we have done, the data cache is the one that grows bigger and faster. The metadata cache is quite small comparatively. The purpose of the proxy is basically to unify all data caches so that memory usage can be kept under control. We have seen that without the proxy, a server starts to have problems with 100-150 simultaneous connections. However, with the proxy we are able to handle, at least, 1000 (we haven't done tests with more connections yet).

On the other side, We have observed a big performance drop when the workload is metadata instensive (or accessing very small files). For data intensive workloads, the performance loss is not so high.

We could consider bypassing the proxy for metadata operations to improve performance. However this would mean having two independent client instances constantly accessing the same files, which may cause some contention between them, and maybe it doesn't compensate the benefit of bypassing the proxy. In any case it may not be a bad idea, at least to try it. I'll think about it and I'll try to do some tests.

**Comment by slowfranklin** [comment]
> > Thinking about it, is there a way to do the metadata operations (including possibly open()) via the proxy, to benefit from (presumably) the shared cache and reduced memory footprint, but do IO (read/write) directly, bypassing the proxy? I imagine that might give you the best of both worlds, sort of...
> 
> If we want to bypass the proxy for some requests, read() and write() are not the best candidates. In some tests we have done, the data cache is the one that grows bigger and faster. 

Hm. Why on earth does it cache data? My idea to avoid the proxy for IO was just to avoid  data copying between processes. For high performance IO the proxy would be a killer otherwise.

> The metadata cache is quite small comparatively. The purpose of the proxy is basically to unify all data caches so that memory usage can be kept under control. We have seen that without the proxy, a server starts to have problems with 100-150 simultaneous connections. However, with the proxy we are able to handle, at least, 1000 (we haven't done tests with more connections yet).

good.

> On the other side, We have observed a big performance drop when the workload is metadata instensive (or accessing very small files).

hm. Do you have any idea why that happens? Shouldn't the proxy provide a (shared) cache for the metadata ops?

> For data intensive workloads, the performance loss is not so high.

For really high performance IO the data copies between processes is going to kill you. :)

**Comment by xhernandez** [comment]
> > If we want to bypass the proxy for some requests, read() and write() are not the best candidates. In some tests we have done, the data cache is the one that grows bigger and faster.
> 
> Hm. Why on earth does it cache data?

For the same reason that the kernel caches data going to the disk. The latency of going to the backend (disk for the kernel, or a remote network node for CephFS) is much higher than the latency of accessing a cache.

> My idea to avoid the proxy for IO was just to avoid data copying between processes. For high performance IO the proxy would be a killer otherwise.

It's a trade-off. Bypassing the proxy provides better performance, but eats the server resources much faster, which ultimately causes failures. So performance is irrelevant when the server dies. If the use case requieres just few connections, the proxy can be avoided to gain performance. It's not mandatory to use it. But if many hundreds or thousands of connections are required, the proxy is necessary, at the cost of some performance.

> > On the other side, We have observed a big performance drop when the workload is metadata instensive (or accessing very small files).
> 
> hm. Do you have any idea why that happens?

The drop in performance happens because each request has more latency because of the additional hop caused by the proxy. When the amount of data transferred is very small, the driving factor for latency becomes the communication latency. For data operations, the latency of transferring it through the network is much higher than the extra communication latency, so the impact is much smaller. Note that the communication between the proxy client and the proxy server happens through a local unix socket, so the bandwidth is very high and the latency quite small compared with other latencies involved in the process.

> Shouldn't the proxy provide a (shared) cache for the metadata ops?

I tested with a metadata cache in the proxy client side, which dramatically improved the performance of metadata operations. However that was just a toy implementation. To implement it in a safe way, it's necessary to correctly handle CephFS' invalidation notifications, and right know I'm not 100% sure of the semanthics of those notifications, so I can't implement them correctly. Without that, the cache could be returning stale data, which is dangerous and could lead to data corruption.

That's probably one of the first things that will be added to the proxy once the CephFS' notification are correctly understood.

> > For data intensive workloads, the performance loss is not so high.
> 
> For really high performance IO the data copies between processes is going to kill you. :)

IMO, there are many other things that need to be improved before the unix socket bandwidth will become a bottleneck.

There are at least two things that could be done to improve performance:

1. Use shared memory between the proxy client and the proxy server. This can avoid huge data copies.
2. Use asynchronous operations to improve parallel access. This way, the throughput will be maximized (so no real performance loss), just adding a bit of latency.

I have these improvements on my mind, but they will be implemented later, once we are sure the proxy works as intended and it's stable.

**Comment by vshankar** [comment]
jenkins retest this please

**Comment by vshankar** [comment]
jenkins retest this please

**Comment by vshankar** [comment]
@spuiuk reached out to me regarding merging this PR.

Can anyone from the list of folks who have reviewed or run test with this change give an approval? I'm looking at the bits that touch builds to ensure nothing stands out there. Once I have an approval, I'm fine to merge this change.

**Comment by vshankar** [comment]
jenkins test make check

**Comment by vshankar** [comment]
jenkins test make check

**Comment by spuiuk** [comment]
The proxy module has been used in the upcoming Red Hat Ceph product as part of Ceph-SMB  and has been tested internally by Red Hat QE. We have been able to go from approximately 150 simultaneous SMB client connections to 1000+ client connections in our tests.

On the back of this testing, I am hoping we can merge this proxy into the Ceph product.

**Comment by vshankar** [comment]
jenkins test make check

**Comment by vshankar** [comment]
@spuiuk There is a pending revert for this change: https://github.com/ceph/ceph/pull/58376#pullrequestreview-2269833816. Once approved, I am willing to merge this.

**Review by spuiuk — APPROVED** [review]

**Comment by vshankar** [comment]
Thx for the approval @spuiuk. Will merge this shortly.

**Review by vshankar — APPROVED** [review]
Approved based on @spuiuk approval for tests runs.

**Pull request merged**
